virtio-devices/src/vsock/unix/muxer.rs

```diff
@@ -480,55 +496,73 @@ impl VsockMuxer {
+
+        // we need to read more to get a `connect ` statement
+        if command.len < CONNECT_WITH_SPACE_LOWER_CASE.len() {
+            return match opt_new_line_position {
+                Some(_) => Err(Error::InvalidPortRequest),
+                None => Err(Error::UnixRead(std::io::ErrorKind::WouldBlock.into())),
+            };
+        }
+
+        // check for both upper and lower case connect statements
+        if !command
+            .buf
+            .starts_with(CONNECT_WITH_SPACE_LOWER_CASE.as_bytes())
+            && !command
+                .buf
+                .starts_with(CONNECT_WITH_SPACE_UPPER_CASE.as_bytes())
+        {
+            return Err(Error::InvalidPortRequest);
+        }
+
+        // we parsed correctly `connect ` but need to wait for `\n`
+        if command.buf.len() == command.len {
+            return Err(Error::InvalidPortRequest);
+        }
+
+        let new_line_position =
+            opt_new_line_position.ok_or(Error::UnixRead(std::io::ErrorKind::WouldBlock.into()))?;
+
+        // we now have the newline, we will treat everything in between as the port
+        let port_string_as_bytes =
+            &command.buf[CONNECT_WITH_SPACE_LOWER_CASE.len()..new_line_position];
+        let port_string =
+            std::str::from_utf8(port_string_as_bytes).map_err(|_| Error::InvalidPortRequest)?;
+
+        port_string
+            .trim()
+            .parse::<u32>()
+            .map_err(|_| Error::InvalidPortRequest)
+    }
+
     /// Parse a host "connect" command, and extract the destination vsock port.
     ///
     fn read_local_stream_port(
-        partial_command_map: &mut HashMap<RawFd, PartiallyReadCommand>,
+        command: &mut PartiallyReadCommand,
         stream: &mut UnixStream,
     ) -> Result<u32> {
-        let command = partial_command_map.entry(stream.as_raw_fd()).or_default();
-
-        // This is the minimum number of bytes that we should be able to read, when parsing a
-        // valid connection request. I.e. `b"connect 0\n".len()`.
-        const MIN_COMMAND_LEN: usize = 10;
-
-        // Bring in the minimum number of bytes that we should be able to read.
-        if command.len < MIN_COMMAND_LEN {
-            command.len += stream
-                .read(&mut command.buf[command.len..MIN_COMMAND_LEN])
-                .map_err(Error::UnixRead)?;
-        }
+        // read only a single byte at a time, this is really inefficient but prevents us
+        // to read past the `\n` character which might swallow actual application data
+        // alternative: the bytes that might have been read beyond `\n` would need
+        // to be sent somehow via `MuxerConnection` prior to reading from `stream` again
+        let read_bytes = stream
+            .read(
+                &mut command.buf[command.len..min(command.len + 1, PartiallyReadCommand::BUF_SIZE)],
+            )
+            .map_err(Error::UnixRead)?;
```

Comment:
Before we could do larger reads until we'd hit the minimum possible command size. Why stop that?

Reply:
I wanted to submit the simplest solution I could come up with first, adding potential optimization+complexity later.

You are right, we can read the first few bytes in one chunk. I have ported this part of the old code, hopefully without an off-by-one error :)